perl-ptest: copy xconfig data from perl source directory

rajendra-desai-ni opened the pull request:
The following perl ptests:

  - dist/ExtUtils-ParseXS/t/001-basic
  - dist/ExtUtils-ParseXS/t/002-more
  - dist/ExtUtils-ParseXS/t/003-usage
  - cpan/ExtUtils-Constant/t/Constant
  - cpan/ExtUtils-MakeMaker/t/02-xsdynamic

are erroring out with:
| /usr/lib/perl/ptest/perl_langinfo.h:8:10: fatal error: xconfig.h: No such file or directory
|     8 | #include "xconfig.h"

xconfig.h contains references to the build host architecture and was removed by commit 2e0f30c46802 ("perl: do not install files that contain build host specific data")

However, it is still included from various other places including these tests, and we are still depending on build host architecture data by including the patches from perl-cross recipe, a dependency to perl recipe.

xconfig.h was added back as a copy step in the commit f90922cdeef5 ("update 5.36.1 -> 5.38.0") but was not added back in perl-ptest include file.

Borrowed the logic from the above commit to make a copy of config.h that is specific to the target architecture. The changes in this commit fixes the test failures.

----------------------------------------------------------

([AB#2990669](https://dev.azure.com/ni/DevCentral/_workitems/edit/2990669/) for reference)

Created this PR as I wanted to get the changes reviewed by the team before I could submit a patch upstream.

chaitu236 commented on meta/recipes-devtools/perl/perl-ptest.inc:
```diff
@@ -42,9 +42,6 @@ do_install_ptest () {
 	 sed -i -e '/Autogenerated starting on/d' ${D}${PTEST_PATH}/lib/unicore/mktables.lst
-
-	 # Remove files with host-specific configuration for building native binaries
-	 rm ${D}${PTEST_PATH}/Makefile.config ${D}${PTEST_PATH}/xconfig.h ${D}${PTEST_PATH}/xconfig.sh
```
My understanding is that the xconfig.h added in copy step in the commit you mention doesn't have build reproducibility problems. Is the xconfig.h we're adding back here the same that got copied in the other commit or is this the original file that had build reproducibility problems?

Also, is adding back `Makefile.config` required?

rajendra-desai-ni replied:
>Do we intend to submit this fix upstream?

Yes, this fix is intended to be submitted upstream.

>My understanding is that the xconfig.h added in copy step in the commit you mention doesn't have build reproducibility problems. Is the xconfig.h we're adding back here the same that got copied in the other commit or is this the original file that had build reproducibility problems?

One of the comment lines in the copy step says:
_"# To make it reproducible let's make it a copy of config.h patch that is specific to the target architecture."_
I am not sure what do they mean by "make it reproducible" and whether this new commit really doesn't have build reproducibility problems or is it still the case.

Also, one of the patches in the perl-cross recipe that patches the cross-compile part of the code has this data:
_"perl-cross assumes the sources use xconfig.h with -DUSE_CROSS_COMPILE. With perl 5.20.0, it's no longer true. Regardless of what mainline perl uses, let's stick with the old xconfig.h way for now."_

This is the reason we wanted to revert the previous changes instead of just copying the config.h data.
@gratian, please correct me if I am wrong here.

>Also, is adding back Makefile.config required?

Ideally, we only need xconfig.h file while installing perl ptest tests and the tests doesn't require the other 2 files. But I wanted to just revert the previous changes and didn't want to cause confusion in the upstream, hence kept it as it is.

chaitu236 replied:
https://docs.yoctoproject.org/5.0.7/test-manual/reproducible-builds.html has info about reproducible builds.

It also has a section on how to test reproducibility for a recipe.

Since the other commit copied config.h to xconfig.h to fix reproducibility problems, we should do the same here if feasible.

The original commit asserts that `Makefile.config` causes build reproducibility problems so I think we shouldn't include it back unless either of following is true
1. Makefile.config doesn't cause build reproducibility problems anymore
1. Makefile.config is required for something else.

rajendra-desai-ni replied:
Made changes as per your suggestions:
- Kept the removal of Makefile.config file (old commit) as it is
- Used similar logic of copying config.h to xconfig.h to avoid reproducibility problems

Please let me know if the new changes are fine.

rajendra-desai-ni replied:
But why does removing files (above mentioned commit changes) require sed replacements?
Are you talking about [f90922c](https://github.com/ni/openembedded-core/commit/f90922cdeef5a6a4b711c5be2156c05bdb20d5b5)? If so, then they are in fact removing build specific data in sed replacements later in the below steps.

chaitu236 replied:
Oh I hadn't noticed that https://github.com/ni/openembedded-core/commit/f90922cdeef5a6a4b711c5be2156c05bdb20d5b5 does add a step to remove build specific data by adding xconfig.h to sed replacements.

So looks like both copying config.h and doing sed replacements are necessary for getting reproducible builds. In which case your fix is good.

chaitu236 reviewed (commented):
Do we intend to submit this fix upstream?

chaitu236 reviewed (approved):
Looks good. Can be submitted upstream.

Probably not required as we're just copying another change that's already upstream but I'm curious if you happened to test (as per https://docs.yoctoproject.org/5.0.7/test-manual/reproducible-builds.html#can-i-test-my-layer-or-recipes) if this keeps the build reproducible.

rajendra-desai-ni committed:
perl-ptest: copy xconfig data from perl source directory

The following perl ptests:

  - dist/ExtUtils-ParseXS/t/001-basic
  - dist/ExtUtils-ParseXS/t/002-more
  - dist/ExtUtils-ParseXS/t/003-usage
  - cpan/ExtUtils-Constant/t/Constant
  - cpan/ExtUtils-MakeMaker/t/02-xsdynamic

are erroring out with:
| /usr/lib/perl/ptest/perl_langinfo.h:8:10: fatal error:
xconfig.h: No such file or directory
|     8 | #include "xconfig.h"

xconfig.h contains references to the build host architecture and was
removed by commit 2e0f30c46802 ("perl: do not install files that contain
build host specific data")

However, it is still included from various other places including these
tests, and we are still depending on build host architecture data by
including the patches from perl-cross recipe, a dependency to perl recipe.

xconfig.h was added back as a copy step in the commit f90922cdeef5
("update 5.36.1 -> 5.38.0") but was not added back in perl-ptest include
file.

Borrowed the logic from the above commit to make a copy of config.h
that is specific to the target architecture. The changes in this commit
fixes the test failures.

Signed-off-by: Rajendra Desai <rajendra.desai@emerson.com>

rajendra-desai-ni commented:
> Looks good. Can be submitted upstream.
> 
> Probably not required as we're just copying another change that's already upstream but I'm curious if you happened to test (as per https://docs.yoctoproject.org/5.0.7/test-manual/reproducible-builds.html#can-i-test-my-layer-or-recipes) if this keeps the build reproducible.

@chaitu236 , The QA Issue that I was getting was because I had added the copy step after build host references removal step due to which some build host specific references in xconfig.h file were left unremoved. I have fixed and tested this, and I am no more getting QA Issue with the new changes. Please let me know about the new changes.
